### DISCUSSION TITLE
Make player Invisible to everyone bar themselves

### DISCUSSION THREAD

**Discussion by zax71** [Minestom/Minestom#1919] [Help]
Hi! I’ve got team that sets players to be partially visible and thus requires them to be invisible. How can I make it so that, in F5, the player is fully visible to themselves and their arm shows, thus setting all players bar themselves invisible?

**Comment by Arektor** [comment]
I'd suggest sending an [Entity Effect packet](https://wiki.vg/Protocol#Entity_Effect) to all of this player's viewers so that they see said player as invisible (partially invisible with the proper team setup).
Also, detect when a new viewer is added, send them that same packet. Detect when a viewer is removed, send them a [Remove Entity Effect packet](https://wiki.vg/Protocol#Remove_Entity_Effect). And when you don't want that player to be invisible, simply send all the player's viewers that remove entity effect packet.

That's at least how I would go about it

**Reply by zax71** [reply]
Hi! Thanks for the help, seams like a good method, but I cannot seam to find out how to send that packet. Most examples on GitHub go something like this:

```java
EntityEffectPacket entityEffectPacket = new EntityEffectPacket();
entityEffectPacket.entityId = entity.getEntityId();
entityEffectPacket.potion = new Potion(PotionEffect.INVISIBILITY, (byte) -1, 0);
 ```

But there must have been some code changes recently, because there is not a constructor for `EntityEffectPacket` that has no variables.

**Reply by Arektor** [reply]
In minestom most packets (if not all) are defined as records - which means their fields are immutable.
There is only two constructors: one with a NetworkBuffer argument used for reading incoming packets, and one with all the packet's fields used for creating (and then sending) said packet.

So in your case, you would just provide the entityId and potion arguments in the constructor.
The factorCodec argument can be null - and should be as of 1.20.1 for any effect that isn't darkness.

**Reply by zax71** [reply]
Thanks for all the help so far! I'm still struggling with this lol...

From what you have said I'd think the below would work, but it has no effect on the players visible to me

```java
private void showPlayers(ParkourPlayer player) {
    MinecraftServer.getTeamManager().getTeam("ghostPlayers").removeMember(player.getUsername());
    player.sendPacketToViewers(new RemoveEntityEffectPacket(player.getEntityId(), PotionEffect.INVISIBILITY));
 }

private void hidePlayers(ParkourPlayer player) {
    MinecraftServer.getTeamManager().getTeam("ghostPlayers").removeMember(player.getUsername());
    player.sendPacketToViewers(new EntityEffectPacket(player.getEntityId(), INVIS, null));
}

private void ghostPlayers(ParkourPlayer player) {
    player.setTeam(MinecraftServer.getTeamManager().getTeam("ghostPlayers"));
    player.sendPacketToViewers(new EntityEffectPacket(player.getEntityId(), INVIS, null));
}
```

```java
// Create the team to make players ghosts
MinecraftServer.getTeamManager().createBuilder("ghostPlayers")
    .seeInvisiblePlayers()
    .updateTeamPacket()
    .build();
```

**Comment by virustotalop** [comment]
This code is a bit old so I'm not sure if it still works but the way I handled it was just by listening to the `PlayerPacketOutEvent` and checking to see if the player in the packet was the one it was being sent to like here:
```java
        if (event.getPacket() instanceof SpawnPlayerPacket entityPacket) {
            if (!entityPacket.playerUuid().equals(event.getPlayer().getUuid())) {
                event.setCancelled(true);
            }
        }
```
For players that already exist in the world you could probably send a remove entity packet and then keep track of when you need to add them back to the world and send a spawn packet back to the clients.
